ethylene-mapper/src/main/java/com/github/steanky/ethylene/mapper/signature/constructor/ConstructorSignature.java

```diff
@@ -0,0 +1,291 @@
+
+import java.lang.ref.Reference;
+import java.lang.ref.SoftReference;
+import java.lang.ref.WeakReference;
+import java.lang.reflect.Constructor;
+import java.lang.reflect.Field;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Parameter;
+import java.util.*;
+
+/**
+ * Signature based off of a particular constructor.
+ */
+public class ConstructorSignature implements Signature {
+    private static final Comparator<? super Field> COMPARATOR = Comparator.comparing(field -> {
+        Order order = field.getAnnotation(Order.class);
+        if (order != null) {
+            return order.value();
+        }
+
+        return 0;
+    });
+
+    private final Token<?> genericReturnType;
+    private final Reference<Class<?>> rawClassReference;
+    private final String rawClassName;
+    private final Reference<Class<?>>[] parameterTypes;
+    private final String[] parameterTypeNames;
+
+    //constructor objects are not retained by the classloader, so they might be garbage collected early
+    //use soft reference to reduce the frequency of this occurrence, and resolve the actual constructor at runtime if it
+    //is necessary to do so
+    private Reference<Constructor<?>> constructorReference;
+    private boolean matchesNames;
+
+    private Collection<Map.Entry<String, Token<?>>> types;
+
+    //similarly to constructors, fields are not tied to the classloader, keep a soft reference and be prepared to
+    //re-create as necessary
+    private Reference<Map<String, Field>> namedFieldsReference = new SoftReference<>(null);
+    private Reference<Field[]> fieldsReference = new SoftReference<>(null);
+
+    @SuppressWarnings("unchecked")
+    public ConstructorSignature(@NotNull Constructor<?> constructor, @NotNull Token<?> genericReturnType) {
+        this.genericReturnType = Objects.requireNonNull(genericReturnType);
+        this.constructorReference = new SoftReference<>(Objects.requireNonNull(constructor));
+
+        Class<?> declaringClass = constructor.getDeclaringClass();
+        this.rawClassReference = new WeakReference<>(declaringClass);
+        this.rawClassName = declaringClass.getTypeName();
+
+        Class<?>[] params = constructor.getParameterTypes();
+        this.parameterTypes = new Reference[params.length];
+        this.parameterTypeNames = new String[params.length];
+        for (int i = 0; i < params.length; i++) {
+            Class<?> referent = params[i];
+            parameterTypes[i] = new WeakReference<>(referent);
+            parameterTypeNames[i] = referent.getTypeName();
+        }
+    }
+
+    private static Map.Entry<String, Token<?>> makeEntry(Parameter parameter, boolean parameterHasName) {
+        Name parameterName = parameter.getAnnotation(Name.class);
+        return Entry.of(parameterHasName ? parameter.getName() : (parameterName != null ? parameterName.value() : null),
+            Token.ofType(parameter.getParameterizedType()));
```

Comment:
wouldn't it make more sense to prefer the Name annotation value over the actual parameter name?